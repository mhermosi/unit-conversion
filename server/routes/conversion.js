const convert = require('convert-units');

const units = {
    'KELVIN':       { type: 'T', code: 'K' }, 
    'CELSIUS':      { type: 'T', code: 'C' }, 
    'FAHRENHEIT':   { type: 'T', code: 'F' }, 
    'RANKINE':      { type: 'T', code: 'R' }, 
    'LITERS':       { type: 'V', code: 'l' }, 
    'TABLESPOONS':  { type: 'V', code: 'Tbs' }, 
    'CUBIC-INCHES': { type: 'V', code: 'in3' }, 
    'CUPS':         { type: 'V', code: 'cup' }, 
    'CUBIC-FEET':   { type: 'V', code: 'ft3' }, 
    'GALLONS':      { type: 'V', code: 'gal' }
};

function verifyConversion(req, res) {
    const { inputNumericalValue, inputUnitOfMeasure, studentResponse, targetUnitOfMeasure} = req.body;

    let from_unit = units[inputUnitOfMeasure.toUpperCase()];
    let to_unit   = units[targetUnitOfMeasure.toUpperCase()];

    let validated = 'invalid';
    let variant   = 'danger';

    console.log('======================================');
    console.log(`Values: from ${inputNumericalValue}, to: ${studentResponse}`);
    from_input = Number(inputNumericalValue);
    to_input   = Number(studentResponse);
    if(from_input !== 'NaN' && to_input !== 'NaN') {
        if(typeof from_unit !== 'undefined' && typeof to_unit  !== 'undefined') {
            console.log(`Converting from ${from_unit.code} to ${to_unit.code} value: ${inputNumericalValue}`);
            /**
             * if the unit type are the same, and the code (Unit) are differenet we calculate the
             * conversion.
             */
            if(from_unit.type == to_unit.type && from_unit.code != to_unit.code)  {
                conversion = Math.round(10 * convert(from_input).from(from_unit.code).to(to_unit.code)) / 10;
                console.log(`Converted value: ${conversion}`);
                if(conversion == to_input) {
                    validated = 'correct';
                    variant   = 'success';
                } else {
                    validated = 'incorrect';
                }
            } 
        }
    } else {
        validated = 'incorrect';
    }

    console.log(`Response: ${validated}`);
    

    res.send({ output: validated, variant: variant });
}

module.exports = { verifyConversion };