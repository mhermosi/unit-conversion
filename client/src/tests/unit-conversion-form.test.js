import React from 'react';
import UnitConversionForm from '../components/unit-conversion-form';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import axiosMock from 'axios';
import conversion from './mock_api';
import cases from './cases';

jest.mock('axios');


test('Load and display form', async () => {
    const { asFragment } = render(<UnitConversionForm  api={conversion} /> );
    expect(asFragment).toMatchSnapshot();
});

describe('Test Scenarios calling handleSubmit', () => {

    it('Test Submit button', async () => {
        const url = '/api/verify-conversion';
        const { container, asFragment } = render(<UnitConversionForm api={conversion} /> );
        expect(output.textContent).toBe('');
        
        axiosMock.get.mockResolvedValueOnce({
            output: 'invalid', variant: 'danger'
        });

        fireEvent.click(screen.getByText('Submit'));

        const output1 = container.querySelector('div[id=output]');
        await waitFor(() => expect(output.textContent).toBe('invalid') );
        expect(asFragment).toMatchSnapshot();
    });

    test.each(cases.test_cases)(
        '%s', 
         async (title, inputNumericalValue, inputUnitOfMeasure, targetUnitOfMeasure, studentResponse, expectedOutput, expectedVariant)  => {

        const url = '/api/verify-conversion';
        const { container, asFragment } = render(<UnitConversionForm api={conversion}/> );
    
        axiosMock.get.mockResolvedValueOnce({
            output: expectedOutput, variant: expectedVariant
        });

        const input1 = container.querySelector('input[id=inputNumericalValue]');
        const input2 = container.querySelector('select[id=inputUnitOfMeasure]'); 
        const input3 = container.querySelector('select[id=targetUnitOfMeasure]');
        const input4 = container.querySelector('input[id=studentResponse]');

        
        fireEvent.change(input1, { target: {value: inputNumericalValue}});
        fireEvent.change(input2, { target: {value: inputUnitOfMeasure}});
        fireEvent.change(input3, { target: {value: targetUnitOfMeasure}});
        fireEvent.change(input4, { target: {value: studentResponse}});
        
        const output = container.querySelector('div[id=output]');
        expect(output.textContent).toBe('');
        
        fireEvent.click(container.querySelector('button[type=submit]'));
        
        //expect(axiosMock.get).toHaveBeenCalledTimes(count++);
        //expect(axiosMock.get).toHaveBeenCalledWith(url);

        await waitFor(() => expect(output.textContent).toBe(expectedOutput) );
        //await expect(output.textContent).toBe(expectedOutput);
        expect(asFragment).toMatchSnapshot();
    });
    
});

