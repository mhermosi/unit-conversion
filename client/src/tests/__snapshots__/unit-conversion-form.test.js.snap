// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load and display form 1`] = `[Function]`;

exports[`Test Scenarios calling handleSubmit Test Submit button 1`] = `[Function]`;

exports[`Test Scenarios calling handleSubmit Test case 1 - correct conversion from Farenheit to Rankine 1`] = `[Function]`;

exports[`Test Scenarios calling handleSubmit Test case 2 - incorrect conversion from Farenheit to Rankine (Student Response wron decimal roundup) 1`] = `[Function]`;

exports[`Test Scenarios calling handleSubmit Test case 3 - incorrect conversion from Kelvin to Fahrenheit (Student Response wrong decimal roundup) 1`] = `[Function]`;

exports[`Test Scenarios calling handleSubmit Test case 4 - correct conversion from Cups to Liters 1`] = `[Function]`;

exports[`Test Scenarios calling handleSubmit Test case 5 - invalid conversion from Gallons to Kelvin 1`] = `[Function]`;

exports[`Test Scenarios calling handleSubmit Test case 6 - incorrect conversion from Fahrenheit to Rankine (student response wrong input) 1`] = `[Function]`;

exports[`Test Scenarios calling handleSubmit Test case 7 - invalid conversion From dog to Celsius (Input Unit of measure wrong input) 1`] = `[Function]`;
